p2p/http/responsewriter.go

```diff
@@ -0,0 +1,146 @@
+package p2phttp
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var bufWriterPool = sync.Pool{
+	New: func() any {
+		return bufio.NewWriterSize(nil, 4<<10)
+	},
+}
+
+var bufReaderPool = sync.Pool{
+	New: func() any {
+		return bufio.NewReaderSize(nil, 4<<10)
+	},
+}
+
+var log = logging.Logger("p2phttp")
+
+var _ http.ResponseWriter = (*httpResponseWriter)(nil)
+
+type httpResponseWriter struct {
+	w                     *bufio.Writer
+	directWriter          io.Writer
+	header                http.Header
+	wroteHeader           bool
+	inferredContentLength bool
+}
+
+// Header implements http.ResponseWriter
+func (w *httpResponseWriter) Header() http.Header {
+	return w.header
+}
+
+// Write implements http.ResponseWriter
+func (w *httpResponseWriter) Write(b []byte) (int, error) {
+	if !w.wroteHeader {
+		if w.header.Get("Content-Type") == "" {
+			contentType := http.DetectContentType(b)
+			w.header.Set("Content-Type", contentType)
+		}
+
+		if w.w.Available() > len(b) {
+			return w.w.Write(b)
+		}
+	}
+
+	// Ran out of buffered space, We should check if we need to write the headers.
+	if !w.wroteHeader {
+		// Be nice for small things
+		if w.header.Get("Content-Length") == "" {
+			w.inferredContentLength = true
+			w.header.Set("Content-Length", strconv.Itoa(len(b)+w.w.Buffered()))
```

Comment:
This would set a wrong content length if `Write` is called multiple times.

Reply:
It doesn't. It's subtle. I added a test to show this.